docs/arch/adr-15.md

```diff
@@ -0,0 +1,38 @@
+# ADR 15: Drawing tools API
+
+June 19, 2019
+
+## Context
+
+The way the drawing tools currently function on Panoptes-Front-End (PFE) have numerous issues including:
+
+- Updating the classification annotation on each touch or pointer event which causes unnecessary re-rendering of the DOM
+- The separation concerns are not clear between components and stores. Multiple components can update the annotation making it hard to debug or add new features to. 
+  - Example: The `MarkingsRenderer` and the `FrameAnnotator` both call change handlers that update the classification annotation? Can the drawing annotation be updated by both or is one solely responsible? It is unclear by reading the code. Why does something named `MarkingsRenderer` update the annotation?
+
+
+## Decision
+
+What we do not want to do:
+- Re-render on every pointer or touch event.
+- update annotation state while drawing is in progress.
+- support more than one drawing task in a step.
+- Use D3.js since it has its own internal data store and it would be complicated to integrate that with a observable stream.
+
+What we do want to do:
+- Have a component, the interactive layer, that manages the interaction with the marks and pointer and touch events.
+  - The interactive layer should not allow events to bubble so the events are encapsulated to just the interaction with the subject. This is to help prevent browser scroll during drawing. An attempted fix on PFE for reference: zooniverse/Panoptes-Front-End#5411
+  - Events will be observed and be streamed via an observable. We will use rx.js to create an observer/observable event stream.
+  - The last state of the event stream will be used to add an annotation to the classification when next or done is clicked.
+- Have a component, the markings renderer, that manages the rendering of previous annotation marks as read only. It will hide them if hide previous marks is toggled.
+- These two components will initially use separate SVG layers that connect to the same stores. Later any duplicated code will be refactored into a higher order component.
+- Have a multi-image subject viewer that will create an interactive layer and markings renderer for each frame. Each interactive layer will initialize its own event stream. We will have to continue to support the configuration of having drawing marks across all frames and of independent marks for each frame. TODO: example of what these mean and how they will be accomplished.
```

Comment:
@eatyourgreens I've added this bullet point describing what we'll do with multi-image subjects. Feel free to edit this file to add your thoughts about it.

Reply:
No problem. A single drawing layer, which filters the annotations by frame number before rendering, might be faster than multiple layers.

Reply:
Multi-image subjects are complicated by the option to show the images side-by-side or in a flipbook. If they're all visible together, then PFE renders a new drawing layer for each one. However, this causes headaches when handling events eg. deleting or moving a mark has to be copied across layers. Opening subtasks gets weird if each copy of a mark has its own task component.

See [Space Warps HSC](https://www.zooniverse.org/projects/aprajita/space-warps-hsc/classify) for a live project that draws on four frames simultaneously with the point tool.

Reply:
In PFE, the drawing layer is also used to handle zoom, pan and rotation transformations, which also gets clunky with projects like Space Warps, where the transformations are not copied across frames.

Reply:
That Space Warp's classify page is crashing for me, so I can't check it out. I'm not sure what's wrong with it.

Reply:
I've updated the description of what we will do to include the two options proposed for multi-image subjects. Is this clear enough for an ADR?

Reply:
I'm not sure what's going on with Space Warps, but the feedback code is crashing the classifier. That link worked when I posted it last week.